### PR TITLE
feat(QZoom): new core component

### PR DESCRIPTION
**Pull request by hawkeye64**

### CODE REVIEW

File: docs/src/pages/vue-components/zoom.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+<doc-example title="Video" file="QZoom/Video" />
+
+### Content
+
+<doc-example title="Content" file="QZoom/Content" />
+
+### Restore on Scroll
+
+In most cases, QZoom turns off any scollbars for a better User Experience. Using the `restore-on-scroll` property allows the scrollbars (if any) to be visible. If the user attempts to scoll, the content will automatically be zoomed out.
```
Comment (jean-moldovan) [inline docs/src/pages/vue-components/zoom.md]: typo `scrollbars`

Comment (jean-moldovan) [reply docs/src/pages/vue-components/zoom.md]: user attempts to `scroll`

File: ui/src/components/zoom/QZoom.json [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  "slots": {
+    "default": {
+      "extends": "default"
+    }
+  },
+
+  "events": {
+    "before-zoom": {
+      "desc": "Trigger when default slotted content is about to be zoomed"
+
```
Comment (jean-moldovan) [inline ui/src/components/zoom/QZoom.json]: empty line

File: docs/src/pages/vue-components/zoom.md (hunk 1 above)
Comment (hawkeye64) [reply docs/src/pages/vue-components/zoom.md]: apparently, my `r` key is broken :)

File: ui/src/components/zoom/QZoom.js [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+      // reset scaling
+      this.setScale(1.0)
+      this.setScaleText(100)
+
+      this.position = this.__getPosition()
+      setTimeout(() => {
+        this.__resetTransition()
+        this.__removeClass(document.body, 'q-zoom__no-scroll')
+        this.$emit('restored')
+      }, 400)
```
Comment (jean-moldovan) [inline ui/src/components/zoom/QZoom.js]: Are these timeout numbers [50, 100, 400] have been dictated by css animations?

File: ui/src/components/zoom/QZoom.js [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    __addClass (el, name) {
+      let arr = el.className.split(' ')
+      // make sure it's not already there
+      if (arr.indexOf(name) === -1) {
+        arr.push(name)
+        el.className = arr.join(' ')
+      }
+    },
+
+    __removeClass (el, name) {
```
Comment (jean-moldovan) [inline ui/src/components/zoom/QZoom.js]: These two can make nice reusable helpers. Not used anywhere else in the app?

File: ui/src/components/zoom/QZoom.js [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+      let position = offset(this.$el)
+      position.left = position.left + 'px'
+      position.top = position.top + 'px'
+      position.width = this.$el.clientWidth + 'px'
+      position.height = this.$el.clientHeight + 'px'
+
+      return position
+    },
+
+    __deepClone (vnodes, createElement) {
```
Comment (jean-moldovan) [inline ui/src/components/zoom/QZoom.js]: Is this a recipe taken from somewhere else? Maybe add a comment.

**Comment by hawkeye64** [comment]
@jean-moldovan Thanks for reviewing for me :heart:

**Pull request merged**